### DISCUSSION TITLE
Multiple Service Call with One State Action

### DISCUSSION THREAD

**Discussion by OzGav** [ExperienceLovelace/ha-floorplan#137] [Q&A]
Thought I would post it here for the benefit of others. @pkozul or @exetico can you have multiple service calls under the one state call. When I do the folliwing it only appears to be doing the image_set

```
            - name: Pool Heating Control
              entities:
                - switch.gpo2_2
              tap_action:
                service: switch.toggle
              state_action:
                - action: call-service
                  service: floorplan.image_set
                  service_data: |
                    >
                    var imageName = '';
                    switch (entity.state) {
                      case 'off':
                        imageName = 'Pool_Heating_Off';
                        break;
                      case 'on': imageName = 'AC_on_heat';
                        break;
                    }
                    return '/local/floorplan/images/' + imageName + '.svg';
                - action: call-service
                  service: floorplan.class_set
                  service_data: 
                    class: '${(entity.state === "on") ? "lamp-on" : (entity.state === "off") ? "lamp-off" : "entity-unavailable"}'

```

**Comment by exetico** [comment]
The state_action are a simpel switch statement `switch (service) {}`. It will actually check for each case, but currently it's `break;`-ing, after the first known action. 

https://github.com/ExperienceLovelace/ha-floorplan/blob/2a9b564a99f200a04d33fe77e8573e4535c6371c/src/components/floorplan/floorplan-element.ts#L1914-L2133

I'm not sure if there's any good reason to do so. @pkozul are the author of most of the code-base, so I'm not 100% into every case. But as pr. my understand, we could just remove the `break;`, and even consider looping through the action-object, if the same action are defined multiple times (multiple image_set in same action, and so on).

I don't do any changes, before we're aligned with @pkozul.

With that said:
[Kindly submit a feature-request](https://github.com/ExperienceLovelace/ha-floorplan/issues/new?assignees=&labels=feature+request%2Cwaiting+on+review&template=feature_request.md&title=%5BFEATURE%5D+), @OzGav, and as always: Thank you, for all the things you're helping with in the community section 😃 !

**Comment by pkozul** [comment]
Hi guys,

Strange, as this behaviour is at the core of floorplan.

The [Multi Floor Example](https://experiencelovelace.github.io/ha-floorplan/docs/example-multi-floor/) has several rules that contain 4 service calls in a sequence, and they all work as expected.

@OzGav Can you please double-check to see if there could be a logical explanation? If you swap the two service calls around, does the `floorplan.class_set` service get called instead?

@exetico The `break`  you mentioned is for breaking out the switch statement rather than the loop that processes each of the service calls.

**Reply by exetico** [reply]
Oh, damn. I was a bit too fast yesterday 😅.

**Comment by OzGav** [comment]
Still not working if I swap it around. I have tried a different CSS class but that made no difference. I normally use class_set to change the style of embedded PNG images. In this instance I am trying to adjust an SVG. Could that be the problem?

**Reply by pkozul** [reply]
@OzGav I wonder if it's a timing thing.

So in your case, the first service call sets the SVG image, and the second service call sets the CSS class on that image.

If  that's the case, then it might just be that the second service call is actually executed before the first service call fully completes, that is, before the actual image has been loaded.

It might be worth us adding a new type of service, i.e. `floorplan.delay` to allow us to wait a certain duration before executing the next service in the sequence. Let's assume we add this new service, you add this new service call in between your two service calls, and specify a wait time of several seconds, and if that results in the style being set successfully, then we know it's a timing thing.

Having said that, we should probably introduce a new option (at the rule level) that allows us to specify whether the rule's services should be run synchronously (each service call needs to complete before the next service call is executed) or in parallel (service calls are executed, without waiting for any of them to complete).

Leave it with me, and I will put something together...

**Reply by OzGav** [reply]
If you definitely think that is it…. However,  tomorrow I will try a few other things such as removing the image_set and trying with a PNG and with cache set to false.

**Reply by pkozul** [reply]
Hi @OzGav

I ran a test using the existing floorplan code, and was able to apply a CSS class to a dynamically-loaded SVG.

Here is what I did...

I added a second service call to the first rule in the [Home](https://experiencelovelace.github.io/ha-floorplan/docs/example-home/) example:

```yaml
    rules:
      - entity: light.garage
        element: light.garage
        state_action:
          - action: call-service
            service: floorplan.image_set
            service_data:
              image: /local/floorplan/examples/home/light_${entity.state}.svg
              cache: true
          - action: call-service
            service: floorplan.class_set
            service_data: test-fill
        tap_action: toggle
```

And added the relevant CSS rule:

```css
.test-fill, .test-fill * {
  fill: green !important;
}
```

Just for completeness, the target element in the SVG where the image gets loaded:

```svg
  <g
     id="light.garage"
     inkscape:label="light.garage">
```

And the result was as expected:

![image](https://user-images.githubusercontent.com/2073827/128577723-5b35f84a-b983-4974-a739-b6c16dec7f99.png)

Can you try a similar CSS rule, to try and fill your entire SVG a certain colour? If it works, then at least we know that floorplan is doing the right thing, and then we can tackle the CSS.

**Reply by OzGav** [reply]
For info I tired by using PNGs and it worked fine. With the SVG I tried this to test as per your above but it made no difference:

```
            - name: Pool Heating Control
              entities:
                - switch.gpo2_2
              tap_action:
                service: switch.toggle
              state_action:
                - service: floorplan.class_set
                  service_data: entity-unavailable
                - service: floorplan.image_set
                  service_data: |
                    >
                    var imageName = '';
                    switch (entity.state) {
                      case 'off':
                        imageName = 'Pool_Heating_Off';
                        break;
                      case 'on': imageName = 'AC_on_heat';
                        break;
                    }
                    return '/local/floorplan/images/' + imageName + '.svg';
```

CSS is:

```
.entity-unavailable, .entity-unavailable * {
  fill: #FF0000 !important;
  opacity: 0.5 !important;
  
}
```

**Reply by exetico** [reply]
That's a crazy bump, with minimal effort m8 😅😁.
